subsys/llext/llext_load.c

```diff
@@ -97,12 +108,36 @@ static int llext_load_elf_data(struct llext_loader *ldr, struct llext *ext)
 
 	ldr->sect_map = llext_alloc(sect_map_sz);
 	if (!ldr->sect_map) {
-		LOG_ERR("Failed to allocate memory for section map, size %zu", sect_map_sz);
+		LOG_ERR("Failed to allocate section map, size %zu", sect_map_sz);
 		return -ENOMEM;
 	}
-
 	memset(ldr->sect_map, 0, sect_map_sz);
-	ext->alloc_size += sect_map_sz;
+
+	ldr->sect_hdrs = llext_peek(ldr, ldr->hdr.e_shoff);
+	if (ldr->sect_hdrs) {
+		ldr->sect_hdrs_on_heap = false;
+	} else {
+		size_t sect_hdrs_sz = ldr->sect_cnt * sizeof(ldr->sect_hdrs[0]);
+
+		ldr->sect_hdrs = llext_alloc(sect_hdrs_sz);
+		if (!ldr->sect_hdrs) {
+			LOG_ERR("Failed to allocate section headers, size %zu", sect_hdrs_sz);
+			return -ENOMEM;
+		}
+		ldr->sect_hdrs_on_heap = true;
+
+		ret = llext_seek(ldr, ldr->hdr.e_shoff);
+		if (ret != 0) {
+			LOG_ERR("Failed to seek for section headers");
```

Comment:
this immediately looks like a memory leak. I think it's better to free all partially allocated resources *inside* the function that allocated them but failed to complete successfully. This just feels more robust to me: either the function succeeds and everything that it had to allocate is allocated, or it failed and returned the system to the state before the call. At the very least I'd add comments, saying that the caller will free...

Reply:
I tend to agree but in this case we *do* cleanup, just one level up in the call stack no?

Reply:
That also applies to almost every other error path in  @this file. :slightly_smiling_face: It was purposefully done to avoid littering the code with `free`s and/or `goto`s, which I find a lot more complicated to verify. My rationale was:
- all pointers are NULL at the beginning of `llext_load` (zeroed out `struct llext`); 
- error out liberally in `do_llext_load` and depending functions, whatever the stage;
- free stuff appropriately once, in `llext_load_cleanup`, according to the result.

Would adding a very big comment describing this be OK?